File: src/lib/solana-rpc.ts
import { config } from "./config";

// Public Solana RPC endpoint
// Note: Public RPC may block browser CORS requests. For production,
// consider using a dedicated RPC provider (Helius, QuickNode, etc.)
const RPC_ENDPOINT = "https://api.mainnet-beta.solana.com";

// SFDP API endpoint
const SFDP_API = "https://api.solana.org/api/community/v1/sfdp_participants";

export interface NetworkComparison {
  totalValidators: number;
  skipRatePercentile: number; // Lower is better (1 = best)
  voteSuccessPercentile: number; // Lower is better
  stakePercentile: number; // Lower is better (1 = most stake)
  networkAvgSkipRate: number;
  networkAvgVoteSuccess: number;
}

export interface SfdpStatus {
  isParticipant: boolean;
  programName?: string;
  status?: string;
  onboardingDate?: string;
}

// API response types
interface RpcResponse<T> {
  result?: T;
  error?: { message: string };
}

interface SfdpParticipant {
  identity?: string;
  vote_account?: string;
  program_name?: string;
  status?: string;
  onboarding_date?: string;
}

// Cache settings
const CACHE_TTL = 5 * 60 * 1000;
let networkCache: { data: NetworkComparison | null; timestamp: number } | null = null;
let sfdpCache: { data: SfdpStatus | null; timestamp: number } | null = null;

/**
 * Make a JSON-RPC call to Solana
 */
async function rpcCall<T>(method: string, params: unknown[]): Promise<T> {
  const response = await fetch(RPC_ENDPOINT, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      jsonrpc: "2.0",
      id: 1,
      method,
      params,
    }),
  });

  if (!response.ok) {
    throw new Error(`RPC error: ${response.status}`);
  }

  const json = (await response.json()) as RpcResponse<T>;
  if (json.error) {
    throw new Error(`RPC error: ${json.error.message}`);
  }

  return json.result as T;
}

/**
 * Fetch network comparison data using getVoteAccounts
 * Calculates percentile rankings for this validator
 */
export async function getNetworkComparison(
  currentSkipRate: number,
  currentVoteSuccess: number,
  currentStake: number
): Promise<NetworkComparison | null> {
  if (networkCache && Date.now() - networkCache.timestamp < CACHE_TTL) {
    return networkCache.data;
  }

  try {
    const result = await rpcCall<{
      current: {
        votePubkey: string;
        activatedStake: number;
        commission: number;
        lastVote: number;
        rootSlot: number;
      }[];
    }>("getVoteAccounts", [{ commitment: "confirmed" }]);

    const validators = result.current ?? [];
    const totalValidators = validators.length;

    if (totalValidators === 0) {
      return null;
    }

    // Estimate percentile based on stake rank
    const stakes = validators.map((v) => v.activatedStake).sort((a, b) => b - a);
    const stakeRank = stakes.findIndex((s) => s <= currentStake * 1_000_000_000) + 1;
    const stakePercentile = Math.round((stakeRank / totalValidators) * 100);

    // Network averages (typical values from reports)
    const networkAvgSkipRate = 0.2;
    const networkAvgVoteSuccess = 99.5;

    // Estimate percentiles based on current values vs network average
    const skipRatePercentile =
      currentSkipRate <= networkAvgSkipRate
        ? Math.round((1 - (currentSkipRate / networkAvgSkipRate) * 0.5) * 50)
        : Math.round(50 + (currentSkipRate / networkAvgSkipRate - 1) * 50);

    const voteSuccessPercentile =
      currentVoteSuccess >= networkAvgVoteSuccess
        ? Math.round((1 - (currentVoteSuccess - networkAvgVoteSuccess) / (100 - networkAvgVoteSuccess)) * 25)
        : Math.round(25 + ((networkAvgVoteSuccess - currentVoteSuccess) / networkAvgVoteSuccess) * 75);

    const data: NetworkComparison = {
      totalValidators,
      skipRatePercentile: Math.max(1, Math.min(100, skipRatePercentile)),
      voteSuccessPercentile: Math.max(1, Math.min(100, voteSuccessPercentile)),
      stakePercentile: Math.max(1, Math.min(100, stakePercentile)),
      networkAvgSkipRate,
      networkAvgVoteSuccess,
    };

    networkCache = { data, timestamp: Date.now() };
    return data;
  } catch (error) {
    console.error("Failed to fetch network comparison:", error);
    return networkCache?.data ?? null;
  }
}

/**
 * Check SFDP participation status
 */
export async function getSfdpStatus(): Promise<SfdpStatus | null> {
  if (sfdpCache && Date.now() - sfdpCache.timestamp < CACHE_TTL) {
    return sfdpCache.data;
  }

  try {
    const response = await fetch(SFDP_API, {
      headers: { Accept: "application/json" },
    });

    if (!response.ok) {
      console.error("SFDP API error:", response.status);
      return null;
    }

    const participants = (await response.json()) as SfdpParticipant[];

    // Search for our identity in the participants list
    const ourEntry = Array.isArray(participants)
      ? participants.find((p) => p.identity === config.identity || p.vote_account === config.voteAccount)
      : null;

    if (!ourEntry) {
      return null;
    }

    const data: SfdpStatus = {
      isParticipant: true,
      programName: ourEntry.program_name ?? "Mainnet Beta",
      status: ourEntry.status ?? "Active",
      onboardingDate: ourEntry.onboarding_date ?? "",
    };

    sfdpCache = { data, timestamp: Date.now() };
    return data;
  } catch (error) {
    console.error("Failed to fetch SFDP status:", error);
    return null;
  }
}

File: crates/bp-web/src/api/solana_rpc.rs
use serde::{Deserialize, Serialize};

#[cfg(feature = "ssr")]
use super::http::post_json_cached;

#[cfg(feature = "ssr")]
const RPC_ENDPOINT: &str = "https://api.mainnet-beta.solana.com";

/// Network comparison stats for a validator
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NetworkComparison {
    pub total_validators: usize,
    pub skip_rate_percentile: u8,
    pub stake_percentile: u8,
}

#[cfg(feature = "ssr")]
#[derive(Serialize)]
struct RpcRequest {
    jsonrpc: &'static str,
    id: u32,
    method: &'static str,
    params: Vec<serde_json::Value>,
}

#[cfg(feature = "ssr")]
#[derive(Deserialize)]
struct RpcResponse {
    result: Option<VoteAccountsResult>,
}

#[cfg(feature = "ssr")]
#[derive(Deserialize)]
struct VoteAccountsResult {
    current: Vec<VoteAccount>,
    #[serde(default)]
    delinquent: Vec<VoteAccount>,
}

#[cfg(feature = "ssr")]
#[derive(Deserialize)]
struct VoteAccount {
    #[serde(rename = "activatedStake")]
    activated_stake: u64,
}

/// Fetch network comparison data using getVoteAccounts
/// Note: Skip rate percentile is estimated using a heuristic based on typical network average
#[cfg(feature = "ssr")]
pub async fn get_network_comparison(current_skip_rate: f64, current_stake: f64) -> Option<NetworkComparison> {
    let request = RpcRequest {
        jsonrpc: "2.0",
        id: 1,
        method: "getVoteAccounts",
        params: vec![serde_json::json!({"commitment": "confirmed"})],
    };

    let body = serde_json::to_string(&request).ok()?;

    // Use cached POST for RPC calls (5 minute TTL)
    let data: RpcResponse = post_json_cached(RPC_ENDPOINT, &body).await?;
    let result = data.result?;

    // Include both current and delinquent validators for accurate network stats
    let mut all_stakes: Vec<u64> = result
        .current
        .iter()
        .chain(result.delinquent.iter())
        .map(|v| v.activated_stake)
        .collect();

    let total_validators = all_stakes.len();

    if total_validators == 0 {
        return None;
    }

    // Sort descending for percentile calculation
    all_stakes.sort_by(|a, b| b.cmp(a));

    let current_stake_lamports = (current_stake * 1_000_000_000.0) as u64;

    // Find rank: position of first stake <= ours, or 0 if we have highest stake
    // Rank is 1-indexed: rank 1 = top validator
    let stake_rank = all_stakes
        .iter()
        .position(|&s| s <= current_stake_lamports)
        .map(|pos| pos + 1) // Convert 0-indexed to 1-indexed
        .unwrap_or(1); // If not found (we have highest), we're rank 1

    let stake_percentile = ((stake_rank as f64 / total_validators as f64) * 100.0).round() as u8;

    // Estimate skip rate percentile based on typical network average
    // NOTE: This is a heuristic - actual percentile would require per-validator skip rate data
    const NETWORK_AVG_SKIP_RATE: f64 = 0.2; // ~20% typical network skip rate
    let skip_rate_percentile = if current_skip_rate <= NETWORK_AVG_SKIP_RATE {
        // Better than average: 1-50 percentile (lower skip = better = lower percentile)
        ((current_skip_rate / NETWORK_AVG_SKIP_RATE) * 50.0).round() as u8
    } else {
        // Worse than average: 50-100 percentile
        (50.0 + ((current_skip_rate - NETWORK_AVG_SKIP_RATE) / NETWORK_AVG_SKIP_RATE) * 50.0).round() as u8
    };

    Some(NetworkComparison {
        total_validators,
        skip_rate_percentile: skip_rate_percentile.clamp(1, 100),
        stake_percentile: stake_percentile.clamp(1, 100),
    })
}
